### PR TITLE
Fix BN+F.forget

### LINKED ISSUE #5555: The behavior of F.forget is not appropriate for BatchNormalization
The combination of `F.forget` and `BatchNormalization` will cause wrong computation result.
`BatchNormalization` link updates its internal statistics (avg_mean and avg_var) every time the forward computation is called during training time, on the other hand, `F.forget` will call forward computation twice.  So the combination of them will lead to wrong statistics result.

Example:
```py
import chainer
import chainer.functions as F
import chainer.links as L

import numpy as np


class Model(chainer.Chain):

    def __init__(self):
        super().__init__()
        with self.init_scope():
            self.bn1 = L.BatchNormalization(5)

    def forward(self, x):
        h = self.bn1(x)
        # h = F.forget(self.bn1, x)
        return h


np.random.seed(666)
mlp = Model()

x = np.random.uniform(size=(2, 5)).astype('f')
y = mlp(x)
F.sum(y).backward()

with chainer.using_config('train', False):
    # test time
    y = mlp(x)
    print(y)
```

Output (without F.forget):
```
variable([[ 0.6917034   0.8193421   0.667606    0.71579635  0.9209751 ]
          [-0.02388343  0.36777574  0.01308255  0.06104017  0.45613596]])
```
Output (with F.forget):
```
variable([[ 0.6842567   0.79659545  0.6600341   0.70522314  0.89286864]
          [-0.05953163  0.32327673 -0.02182698  0.02312531  0.4058005 ]])
```

This unmatched behavior is not described in document.

### PR DESCRIPTION
**Pull request by ir5**
Fix #5555 by simply adding new global config.

TODO:
- [x] Write test
- [x] Fix F.BatchReNormalization as well?
- [x] Write document somewhere

### CODE REVIEW

**Comment by kmaehashi** [comment]
Jenkins, test this please.

File: chainer/__init__.py [hunk 1]
```diff
@@ -169,6 +169,7 @@ def is_arrays_compatible(arrays):
     raise TypeError('incorrect dtype name in CHAINER_DTYPE: "{}". '
                     'Only float16/32/64 are allowed.'.format(_chainer_dtype))
 global_config.dtype = numpy.dtype(_chainer_dtype)
+global_config.recompute = False
```
Comment (kmaehashi) [inline chainer/__init__.py]: * The configuration key `recompute` sounds too general and a bit confusing to users. How about naming it like `in_recomputation`?
* Could you add the description like this to https://docs.chainer.org/en/latest/reference/configuration.html#configuration-keys?

```rst
This flag is automatically set by :func:`chainer.functions.forget` and not intended to be changed by users.
You can use this flag when implementing your own Link to avoid updating the internal states during recomputation done by `F.forget`.
See the documentation of :func:`chainer.functions.forget` for details.
```

Comment (ir5) [reply chainer/__init__.py]: Thank you for review. I fixed code.

**Comment by ir5** [comment]
I fixed the problems, so this PR is ready for review.

**Comment by ir5** [comment]
OK, I'll check.

**Comment by ir5** [comment]
I fixed when `self.N` is incremented.
However, it seems that this problem occurs only when `BatchNormalization` takes `finetune=True` argument in `**kwargs`, which we cannot pass to the arguments of `F.forget`. Further, current `F.forget` does not take boolean arguments. So, while I think I have fixed the problem, I didn't write test for it.

**Comment by toslunar** [comment]
Could you test it with something like `F.forget(functools.partial(self.link, finetune=True), x)`?

**Comment by ir5** [comment]
Sounds nice. I added test and fixed bug in `BatchRenormalization`.

**Comment by ir5** [comment]
Why Travis CI build failed? It indicates flake8 error at strange point...

**Comment by toslunar** [comment]
I restarted Travis after #5764 is merged.

Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
After discussion, concluded that this issue should be rather considered as bug instead of enhancement.
